RFC-0030-Consolidate-Elastic-and-TorchX.md

```diff
@@ -0,0 +1,321 @@
+An interesting observation is that despite the amount of fragmentation in the tooling for PyTorch,
+when it comes to process launchers, most of these tools either directly invoke or are compatible with
+`torchrun`. This is the case with [Accelerate](https://huggingface.co/blog/pytorch-ddp-accelerate-transformers),
+[DeepSpeed](https://github.com/microsoft/DeepSpeed/blob/master/deepspeed/launcher/launch.py#L29), 
+[Megatron-LM](https://github.com/NVIDIA/Megatron-LM/blob/main/examples/run_text_generation_server_345M.sh#L17),
+and [SageMaker](https://sagemaker.readthedocs.io/en/stable/frameworks/pytorch/using_pytorch.html#launching-a-distributed-training-job).
+So there is something to be said about the acceptance and adoption of standards when it belongs in 
+the core pytorch library versus it being offered as a separate and optional install. 
+Having TorchX specs/APIs in core encourages the community to adopt and extend PyTorch-defined infrastructure
+standards. For the user, this means better less fragmentation, better portability and uniform UX. 
+For PyTorch, this means being able to leverage and dictate its runtime requirements and environment.
+ 
+
+## **Proposed Implementation**
+
+**TL;DR - Upstream TorchX (specs and sensible default impls only) to PyTorch under a new sub-module called `torch.x` and pull TorchElastic out of
+`torch.distributed` and merge it with `torch.x`**
+
+The diagram below depicts the proposed module move + upstream (optional merge of the `multiprocessing` module not shown).
+
+![`torch.x` Before and After](RFC-0030-assets/modules_before_after.png)
+
+**The following changes are proposed (refer to diagram above):**
+
+1. Move `torch.distribted.elastic.*` &rarr; `torch.x.elastic.*`
+2. Upstream `torchx.*` &rarr; `torch.x.*`. Here we have three options:
+     1. (Option 1) Upstream specs only:
+          1. `torchx.specs` (the interfaces and APIs)
+          2. Default implementations of said specs
+          3. Other implementations would remain in TorchX repo
+          4. PROS: Upstreams the minimal set of functionalities to have torch benefit from the topics mentioned in the [Motivation](#motivation) section 
+          5. CONS: 
+              1. Effectively splits the TorchX repo in two hence makes maintenance and CI/CD (testing) more complex
+              2. No built-in support for launching remote jobs
+     2. **[ PREFERRED ] (Option 2) Option 1 + Kubernetes and SLURM support**
+          1. PROS: Keeps the minimalistic approach of Option 1 while providing out-of-the-box for CSP agnostic remote scheduler support
+          2. CONS: TorchX still split into two repos
+         
+        > NOTE: The rest of the doc assumes Option 2 as this is the recommended option
+     3. (Option 3) Upstream everything in TorchX:
+          1. PROS: Makes maintenance and CI/CD simple since changes to the specs can be tested for downstream BC easily
+          2. CONS: bloat-torch
+3. Merge the functionalities of `torchx` CLI to
```

Comment:
Consolidating around `torchrun` sounds like a good idea. My understanding from the above is that:
* `torchrun` would now take a scheduler arg, and users can launch on Slurm/Kubernetes out of the box with just the PyTorch package. 
* Other schedulers will be implemented in the `torchx` repo, and users would need to install torchx + do some registration to use those schedulers.
* Runtime libraries like artifact tracking will remain in `torchx`, as will non-runtime stuff like pre-defined components, the pipelines API, etc.

With this approach what will be the framing for the torchx repo? It may come across as a set of disparate pieces that might be useful during training.

Reply:
> `torchrun` would now take a scheduler arg, and users can launch on Slurm/Kubernetes out of the box with just the PyTorch package.

Yes exactly. There's a bit of tricky BC to deal with in terms of CLI arguments (`torchrun` and `torchx` CLIs currently have completely different args - `torchx` focuses more on actions that can be done at the job level whereas `torchrun` focuses on a "single action": launching multiple procs on the local node).

> Other schedulers will be implemented in the torchx repo, and users would need to install torchx + do some registration to use those schedulers.

The idea is to upstream the `torchx.specs.*` (iface and APIs) to `torch.x.*` and optionally (but strongly encouraged) upstream 1 non-trivial scheduler integration. IMO SLURM is a good candidate since it has no "client" dependencies (e.g. `torchx.schedulers.kubernetes` uses the [`kubernetes`](https://pypi.org/project/kubernetes/) python client). The existing schedulers (k8s, aws-batch, gcp-batch, azure-batch, ray) should stay in the torchx repo as an optional plugin.

As for registration of the schedulers, this is already possible in torchx by design (see [Registering Custom Schedulers](https://pytorch.org/torchx/latest/advanced.html#registering-custom-schedulers)). In fact this is how we currently make only the internal schedulers visible to Meta-internal uses.


> Runtime libraries like artifact tracking will remain in torchx, as will non-runtime stuff like pre-defined components, the pipelines API, etc.

Yes for now. `torchx.runtime.*` (including artifact tracking) APIs are currently not stable therefore, I think should be excluded from the upstream. The pipeline integrations should be deprecated (since no one seems to be using it). As for the pre-def components, I think its useful to have `python` and `ddp` (probably should be renamed to `spmd`) in the main repo since those are the two most used launch topologies. They would serve two purposes: 

#. As a vanilla/basic quick start components to launch test and PTD jobs
#. As reference implementations to further customize those components to fit your exact needs/infra.

> With this approach what will be the framing for the torchx repo? It may come across as a set of disparate pieces that might be useful during training.

IMHO the TorchX repo should serve as a place to hold commonly used plugins (e.g. widely used scheduler impls, most commonly used resources (host types), etc).